File: demos/05-cli-sdk/02-sdk/02-custom-tools/sdk-demos-solution/weather-agent.ts
import { CopilotClient, defineTool, SessionEvent } from "@github/copilot-sdk";

// Define a custom tool that Copilot can call automatically
const getWeather = defineTool("get_weather", {
  description: "Get the current weather for a city",
  skipPermission: true,
  parameters: {
    type: "object",
    properties: {
      city: {
        type: "string",
        description: "The city name to get weather for",
      },
    },
    required: ["city"],
  },
  handler: async (args: { city: string }) => {
    // In a real app, this would call a weather API
    const conditions = ["sunny", "cloudy", "rainy", "partly cloudy"];
    const temp = Math.floor(Math.random() * 30) + 50;
    const condition = conditions[Math.floor(Math.random() * conditions.length)];

    return {
      city: args.city,
      temperature: `${temp}°F`,
      condition,
      timestamp: new Date().toISOString(),
    };
  },
});

async function main() {
  try {
    const client = new CopilotClient();
    const session = await client.createSession({
      model: "gpt-5-mini",
      streaming: true,
      tools: [getWeather],
    });

    // Stream responses as they arrive
    session.on((event: SessionEvent) => {
      if (event.type === "assistant.message_delta") {
        process.stdout.write(event.data.deltaContent);
      }
      if (event.type === "session.idle") {
        console.log("\n");
      }
    });

    // Send a prompt that requires the agent to use tools
    await session.sendAndWait({
      prompt: "What's the weather like in Seattle and Tokyo? Give me the temperatures.",
    });

    await client.stop();
    process.exit(0);
  } catch (error) {
    console.error("Error:", error);
    process.exit(1);
  }
}

main();

File: demos/05-cli-sdk/02-sdk/02-custom-tools/custom-tools-solution-py/weather_agent.py
import asyncio
import random
from datetime import datetime, timezone

from copilot import CopilotClient, define_tool
from copilot.session_events import AssistantMessageDeltaData, SessionIdleData
from pydantic import BaseModel, Field


class GetWeatherParams(BaseModel):
    city: str = Field(description="The city name to get weather for")


@define_tool("get_weather", description="Get the current weather for a city", skip_permission=True)
async def get_weather(params: GetWeatherParams) -> dict:
    conditions = ["sunny", "cloudy", "rainy", "partly cloudy"]
    temp = random.randint(50, 79)
    condition = random.choice(conditions)

    return {
        "city": params.city,
        "temperature": f"{temp}°F",
        "condition": condition,
        "timestamp": datetime.now(timezone.utc).isoformat(),
    }


def on_event(event):
    match event.data:
        case AssistantMessageDeltaData() as data:
            print(data.delta_content, end="", flush=True)
        case SessionIdleData():
            print("\n")


async def main():
    client = CopilotClient()
    session = await client.create_session(
        model="gpt-5-mini",
        streaming=True,
        tools=[get_weather],
    )

    session.on(on_event)

    await session.send_and_wait(
        "What's the weather like in Seattle and Tokyo? Give me the temperatures."
    )

    await client.stop()


asyncio.run(main())
